src/jit/hwintrinsiccodegenxarch.cpp

```diff
@@ -544,6 +490,74 @@ void CodeGen::genHWIntrinsic_R_R_RM_I(GenTreeHWIntrinsic* node, instruction ins)
     }
 }
 
+// genHWIntrinsicJumpTableFallback : generate the jump-table fallback for imm-intrinsics
+//                       with non-constant argument
+//
+// Arguments:
+//    intrinsic      - intrinsic ID
+//    nonConstImmReg - the register contains non-constant imm8 argument
+//    baseReg        - a register for the start of the switch table
+//    offsReg        - a register for the offset into the switch table
+//    emitSwCase     - the lambda to generate siwtch-case
+//
+// Return Value:
+//    generate the jump-table fallback for imm-intrinsics with non-constant argument.
+// Note:
+//    This function can be used for all imm-intrinsics (whether full-range or not),
+//    The compiler front-end (i.e. importer) is responsible to insert a range-check IR
+//    (GT_HW_INTRINSIC_CHK) for imm8 argument, so this function does not need to do range-check.
+//
+template <typename HWIntrinsicSwitchCaseBody>
+void CodeGen::genHWIntrinsicJumpTableFallback(NamedIntrinsic            intrinsic,
+                                              regNumber                 nonConstImmReg,
+                                              regNumber                 baseReg,
+                                              regNumber                 offsReg,
+                                              HWIntrinsicSwitchCaseBody emitSwCase)
+{
+    assert(nonConstImmReg != REG_NA);
+    emitter* emit = getEmitter();
+
+    const unsigned maxByte = (unsigned)Compiler::immUpperBoundOfHWIntrinsic(intrinsic) + 1;
+    assert(maxByte <= 256);
+    BasicBlock* jmpTable[256];
+
+    unsigned jmpTableBase = emit->emitBBTableDataGenBeg(maxByte, true);
+    unsigned jmpTableOffs = 0;
+
+    // Emit the jump table
+    for (unsigned i = 0; i < maxByte; i++)
+    {
+        jmpTable[i] = genCreateTempLabel();
+        emit->emitDataGenData(i, jmpTable[i]);
+    }
+
+    emit->emitDataGenEnd();
+
+    // Compute and jump to the appropriate offset in the switch table
+    emit->emitIns_R_C(INS_lea, emitTypeSize(TYP_I_IMPL), offsReg, compiler->eeFindJitDataOffs(jmpTableBase), 0);
+
+    emit->emitIns_R_ARX(INS_mov, EA_4BYTE, offsReg, offsReg, nonConstImmReg, 4, 0);
+    emit->emitIns_R_L(INS_lea, EA_PTR_DSP_RELOC, compiler->fgFirstBB, baseReg);
+    emit->emitIns_R_R(INS_add, EA_PTRSIZE, offsReg, baseReg);
+    emit->emitIns_R(INS_i_jmp, emitTypeSize(TYP_I_IMPL), offsReg);
+
+    // Emit the switch table entries
+
+    BasicBlock* switchTableBeg = genCreateTempLabel();
+    BasicBlock* switchTableEnd = genCreateTempLabel();
+
+    genDefineTempLabel(switchTableBeg);
+
+    for (unsigned i = 0; i < maxByte; i++)
+    {
+        genDefineTempLabel(jmpTable[i]);
+        emitSwCase(i);
```

Comment:
@tannergooding @mikedn Changed `emitSwCase` to accept `unsigned`, so that this function only uses `unsigned` variables.